### ISSUE TITLE
[feature]: introduce df-vlm-qa-v1.0 format

### ISSUE THREAD

**Issue by ermu2001** [NVIDIA-TAO/tao-daft#72]
# Add `df-vlm-qa-v1.0` — HITL bounding-box correction exchange format

**Labels:** `kind:feature`, `kind:api-change`, `area:core`, `area:cli`, `area:docs`

## Summary

Register a fourth format, **`df-vlm-qa-v1.0`**, in the DAFT format registry, with its
validator and converters in both directions between it and `tao-vl-reason-v1.0`.

`df-vlm-qa-v1.0` is the hand-off contract between an automated VLM annotation pipeline
and a human-in-the-loop correction stage. It carries VLM QA items *and* the object tracks
those items reference, so an annotator can correct text and boxes in one pass. The format
is in production use today with no schema under version control and no validator.

It is **bidirectional**: one schema describes both the file sent out for correction and
the file handed back.

## The format

One JSON document per video clip. A dataset is a batch directory of such documents plus
the referenced media.

```jsonc
{
  "format": "df-vlm-qa-v1.0",
  "metadata": { "type": "annotation", "date": "2026-08-04", "license": "CC-BY-4.0" },

  "video_id": "source_a/20250605/clips/7ec562a5-….mp4",  // path under the batch media root
  "video_url": "s3://…/batch_001/7ec562a5-….mp4",        // optional; see §4
  "video_sha256": "1859d376…",                           // pins the exact bytes annotated

  "tracking_meta": {                    // declares the ANNOTATION GRID — see below
    "time_unit": "frame_index",         // or "timestamp_s"
    "box_order": "t_x1_y1_x2_y2",       // const — one order, no switch
    "coordinate_space": "pixel",        // or "normalized"
    "width": 1920, "height": 1080,
    "source_fps": 30.0, "frame_count": 300, "duration_s": 10.0,
    "sample_fps": 7.5,
    "source": "sam3:sam3.pt"
  },

  "tracking": [                         // geometry only — no labels, no descriptions
    { "track_id": "person_T001",
      "track": [[120, 1363, 87, 1415, 160], [124, 1359, 87, 1414, 150]] }
  ],

  "sub_tasks": [                        // the tao-vl-reason-v1.0 item shape, plus markers
    { "task_type": "tracking_description",
      "question": "Describe the tracking <track>person_T001</track>.",
      "answer": "man in a white short-sleeve T-shirt <track>person_T001</track>" }
  ]
}
```

Required: `format`, `metadata`, `video_id`, `sub_tasks`. `tracking_meta` is required
whenever `tracking` is non-empty. `additionalProperties: false` at the top level and
inside `tracking[*]` and `sub_tasks[*]`; `metadata` is open.

### The annotation grid

`tracking_meta` declares the set of times the annotation covers rather than leaving it to
be inferred. Grid spacing is `source_fps / sample_fps` source frames from frame 0, for
`ceil(frame_count / spacing)` samples; under `timestamp_s`, spacing is `1 / sample_fps`
seconds across `duration_s`. Every row's time key must land on that grid.

`tracking` is sparse in two senses that must not be confused:

- rows exist only at grid times — *not* a statement about the world;
- within the grid, a track may have gaps — **a grid time with no row asserts the object
  was not visible then.**

Only the second is a claim, and in a corrected hand-back file it is authoritative: the
annotator was shown that grid time and saw nothing. Dropping a row for any reason *other*
than absence (low confidence, a failed frame, truncated output) silently asserts
something false. The grid is declared rather than inferred because pooling the time keys
of every track only recovers it when some track is boxed at every grid time, and fails
silently on exactly the stretches where nothing is boxed at all.

### 1. Object references — `<track>{track_id}</track>`

Tracks are cross-referenced from QA text by wrapping the id inline in `question`,
`answer`, and `reasoning`. These markers are the single source of truth for which tracks
a sub-task references; there is no separate id list to keep in sync. The UI renders them
as clickable, colour-matched pills.

`track_id` matches `^[A-Za-z0-9][A-Za-z0-9_.:-]*$` and must not contain `<`, `>`, or
whitespace.

### 2. Timestamps — seconds with an `s` suffix, untagged

Timestamps are written inline as a seconds value carrying a literal `s` suffix, measured
from the start of the clip, with no surrounding tag.

```
… the forklift stops at 7.50s and reverses between 12s and 14.50s.
```

Matches `^\d+(\.\d{1,2})?s$`. The integer part takes as many digits as the clip needs and
is not zero-padded; the fractional part is at most two digits. A range is two independent
timestamps, not a single spanning token. The suffix is what separates a timestamp from
any other number in the prose, so consumers can pattern-match without a tag.

Colon-separated forms (`00:07.50`, `01:23`, `00:05:30`), bare numbers (`7.50`), and
spelled-out units (`7.5 seconds`) are **not** valid and are reported by the validator.

### 3. Non-correctable sub-tasks — `<SKIP>`

If the content of *any* field of a sub-task (`question`, `answer`, or `reasoning`) begins
with `<SKIP>` — case-insensitive — that sub-task is **not correctable**. Annotation
tooling must skip it, and it must be handed back byte-identical. This lets a producer ship
an item through a batch without asking a human to correct it.

### 4. `video_url` — optional

`video_id` is a path relative to a batch media root that the file itself never names. The
annotation UI is a browser component that fetches media client-side, so it needs a
per-item URI it can presign. Producers that know the destination at upload time stamp it
in; producers that do not omit it and let the consumer join `video_id` to an out-of-band
batch root. It is optional because it is not knowable at inference time — it is stamped at
the upload step. When present it must be an `s3://` URI.

### 5. `tracking[*]` is geometry-only

A track carries no label or appearance description of its own. What the object looks like
is read from the `tracking_description` sub-task whose question references
`<track>{track_id}</track>` — that is how an annotator identifies which object a bare
`track_id` refers to. A track therefore needs such a sub-task to be correctable.

### 6. `sub_tasks[*]` is closed

`additionalProperties: false`. QA verdicts and review state belong in the annotation
tool's own storage, not on the exchange document. Review metadata that needs to travel
later should arrive as named optional fields in a `v1.1`. `metadata` stays open, which is
where free-form producer data goes.

### Format package — `src/nvidia_tao_daft/formats/df-vlm-qa-v1.0/`

- [ ] `schemas/df_vlm_qa.schema.json` — Draft-7, plus the optional `video_url` property
- [ ] `README.md` — status badge, shape, design notes, validation invocation
- [ ] `specs/directory-structure.md` — batch layout, media root, `video_url` vs
      `video_id` resolution
- [ ] `specs/schema-reference.md` — field-by-field tables, one `### <type>` section per
      `metadata.type` const (enforced by `tests/test_schema_doc_consistency.py`); carries
      the documented `task_type` list (`bcq`, `bcq_openended`, `mcq`, `mcq_openended`,
      `open_qa`, `causal_linkage`, `scene_description`, `temporal_description`,
      `temporal_localization`, `video_summarization`, `tracking_description`,
      `grounded_spatial_temporal_description`)
- [ ] Status badge: 🚧 **Active**

### Validator — `src/nvidia_tao_daft/validators/df_vlm_qa_v1_0/`

- [ ] `DfVlmQaV1_0Validator(BaseValidator)`, `format = "df-vlm-qa-v1.0"`, re-exported from
      `validators/__init__.py`, wiring `tao-daft validate df-vlm-qa-v1.0`
- [ ] Checks per the table below, honouring the existing permissive/strict split

### Converters — both directions

- [ ] `pairs/df_vlm_qa_v1_0_to_tao_vl_reason_v1_0.py` —
      `DfVlmQaV1_0ToTaoVlReasonV1_0Converter(BaseConverter)`
- [ ] `pairs/tao_vl_reason_v1_0_to_df_vlm_qa_v1_0.py` —
      `TaoVlReasonV1_0ToDfVlmQaV1_0Converter(BaseConverter)`
- [ ] Both re-exported from `converters/__init__.py` to trigger auto-registration

### Docs, examples, tests

- [ ] Registry rows in `formats/README.md`, `validators/README.md`, and **two** rows in
      `converters/README.md`; version-field row in `formats/versioning.md`
      (`df-vlm-qa-v1.0` declares via `"format"`)
- [ ] `converters/README.md` states "Conversion is uni-directional: data-centric →
      training-centric". This pair is the first exception; restate it as a per-pair
      property
- [ ] `examples/datasets/df-vlm-qa-v1.0/` — a small batch exercising a full-length track,
      a track with a mid-clip absence gap, a `<SKIP>`ped sub-task, `7.50s`-style
      timestamps
- [ ] Tests and valid/invalid fixtures for the validator and both converters, plus a
      forward→reverse round-trip asserting exactly the loss documented under Reverse and
      no more

## Validator checks beyond JSON Schema

JSON Schema catches shape. None of the checks below are expressible in Draft-7.

| Check | Level | Rationale |
|---|---|---|
| Grid spacing `source_fps / sample_fps` is a whole number | error | A fractional spacing leaves the grid undefined |
| Every row's time key lands on the declared grid | error | Off-grid rows get snapped by the UI; the snap is only safe while it stays one-to-one |
| No two rows of the same track share a grid time | error | The case where snapping silently drops the earlier row |
| Rows ascending by time key | error | Schema says ordered; nothing enforces it |
| `tracking_meta` present when boxes were sought, even if `tracking` is empty | warning | An empty-tracking item still needs its grid declared |
| `x1 < x2` and `y1 < y2` | error | Catches a `t_x1_x2_y1_y2` file read as `t_x1_y1_x2_y2`; the schema's `const` box_order cannot catch a mislabelled file |
| Box coordinates within frame (`width`/`height`, or `[0,1]` normalized) | warning | Edge-clipped boxes are legitimate; wildly out-of-frame ones are a coordinate-space bug |
| `track_id` unique within a clip | error | Ambiguous `<track>` resolution otherwise |
| Every `<track>X</track>` resolves to a track in `tracking` | error | Unresolvable object reference |
| Every track in `tracking` referenced by at least one sub-task | warning | An unreferenced box gives the annotator no way to tell what it is |
| Every track has a `tracking_description` sub-task | warning | A track without one is not correctable |
| `<track>` markers well-formed, closed, unnested; inner value matches the `track_id` pattern | error | A malformed marker breaks box linking silently |
| Time-like strings in sub-task text are canonical `SS...SSS.ss` + `s` (`7.50s`) | warning | Bare numbers (`7.50`) and spelled-out units ("7.5 seconds") defeat the consumer's pattern match |
| Colon-separated timestamps (`MM:SS[.ss]`, `HH:MM:SS`) in sub-task text | warning | The superseded form; a consumer reading `01:23` as seconds is off by 83× |
| `<SKIP>` appears at the start of a field, not mid-string | warning | Mid-string almost certainly means it was meant as a prefix |
| Count of `<SKIP>`ped sub-tasks reported in the summary | info | Makes non-correctable volume visible before a batch ships |
| `task_type` outside the documented set | warning | Free-form by design, but a typo silently creates a new partition file downstream (`open_q.json` alongside `open_qa.json`) |
| `video_id` resolves under the media root | warning permissive / error strict | Matches the existing media check on other formats |
| `video_sha256` matches the file on disk | error | The consuming browser cannot hash the file; DAFT is the ingest point that can |
| Media duration vs `frame_count`/`source_fps` (or `duration_s`), 0.5 s tolerance | warning | A re-cut clip shifts every box |
| `video_url`, when present, is a well-formed `s3://` URI | error | A malformed URI fails at presign time, deep in the UI |

Checks needing the media file are skipped with a warning when the media is not reachable,
consistent with the other validators.

## Converter behaviour

Two pair modules. `BaseConverter` is keyed on `(source_format, target_format)` and the CLI
groups by `source_format`, so the reverse pair needs no framework change. The two
directions are not inverses: forward discards geometry, and reverse cannot invent it back.

### Forward — `df-vlm-qa-v1.0` → `tao-vl-reason-v1.0`

`tao-daft convert df-vlm-qa-v1.0 tao-vl-reason-v1.0 --path <batch>/ --output <dst>/`

Turns a corrected batch into training files.

- One `tao-vl-reason-v1.0` file per `task_type` — `open_qa.json`,
  `scene_description.json`, … — carrying `metadata.task`, matching how
  `tao-vl-reason-v1.0` datasets are already partitioned
- Each sub-task becomes one item: `video_id`, `question`, `answer`, and `reasoning` when
  non-empty. `media_root: null`
- `item_index` is `<clip-stem>:<sub_task index>`, giving every training item a route back
  to the exact sub-task of the exact clip. `video_url` carries through when present —
  `tao-vl-reason-v1.0` items are `additionalProperties: true`
- `<SKIP>`ped sub-tasks are dropped and counted in `ConversionResult.samples_skipped`;
  they were never human-corrected
- `tracking` is dropped — `tao-vl-reason-v1.0` has nowhere to put boxes. Lossy by
  construction, and the converter warns. The geometry stays in the `df-vlm-qa-v1.0` files
- `metadata.date` and `metadata.license` carry through; a non-uniform `metadata` across
  the batch is an error

### Reverse — `tao-vl-reason-v1.0` → `df-vlm-qa-v1.0`

`tao-daft convert tao-vl-reason-v1.0 df-vlm-qa-v1.0 --path <dst>/ --output <batch>/`

Seeds a correction batch from an existing training set, so QA items that never went
through human review can.

- Regroups by clip, inverting the forward per-`task_type` partitioning: reads every
  `tao-vl-reason-v1.0` file under `--path`, emits one document per distinct `video_id`.
  A directory-level operation, not a file-level one
- `item_index` is the join key — its `<clip-stem>:<sub_task index>` form recovers both the
  grouping and the original `sub_tasks` ordering. When absent, the converter falls back to
  file-then-array order and warns
- `question` / `answer` / `reasoning` map back into a `sub_tasks` entry. `metadata.task`
  becomes `task_type` when set, else `--default-task-type` is required
- `metadata.date` / `metadata.license` carry through; `video_url` is picked back up from
  the item if it rode along

What cannot be reconstructed:

| Field | Result | Why |
|---|---|---|
| `tracking` | `[]` | No geometry survives the forward direction |
| `tracking_meta` | omitted | The grid is a property of a tracker run, not of QA text |
| `video_sha256` | recomputed from media when reachable, else omitted | Guessing it would be worse than omitting it |
| `<SKIP>`ped sub-tasks | not restored | `item_index` reveals the gap but cannot fill it |

Empty `tracking` with no `tracking_meta` is schema-valid, so a seeded batch validates — but
trips the "`tracking_meta` present when boxes were sought" warning, correctly: a batch
destined for box correction still needs a tracker run to supply the grid before it ships.

- [ ] `--geometry-from <batch>/` — optional flag pointing at an existing
      `df-vlm-qa-v1.0` batch, re-attaching `tracking`, `tracking_meta`, and
      `video_sha256` by `video_id`. Turns the reverse direction into a true round-trip

## Open questions

1. Marker handling in the forward converter — strip `<track>person_T001</track>` to
   `person_T001`, drop it, or keep it? Suggest `--markers {keep,strip,drop}`, default
   `keep`.
2. Should `tracking_description` sub-tasks convert at all? They are annotation
   scaffolding rather than a QA task worth training on. Suggest converting by default,
   easy to exclude.
3. `grounded_spatial_temporal_description` appears in the delivered example but not in the
   documented task list. Confirm before baking it in.
4. Is `--geometry-from` sufficient for round-trip, or should the forward converter
   preserve geometry itself (sidecar, or riding along on items)?
5. Reverse `task_type` — `--default-task-type`, or infer from the filename stem the
   forward direction writes?

## References

- The schema and reference item in production use — `df_vlm_qa_v1_0.schema.json` and
  `example_7ec562a5.df-vlm-qa-v1.0.json` — to be folded into this repo
- One-off implementations to replace: `convert_to_df_vlm_qa_v1_0.py` (producer),
  `convert_from_df_to_tao_vl_reason.py` (delivery)
